### ISSUE TITLE
cli:  add support for JSON formatted output

### ISSUE THREAD

**Issue by warthog618** [warthog618/gpiocdev-rs#4]
Ass suggested in pr #3, add a `--json` option to the cli commands to format the output as JSON.
For most commands this will take the form of a complete JSON document.
For edges and notify it will be a sequence of JSON fragments, each describing an event, possibly with the option to squash each event onto one line rather than using the pretty-print format.

**Comment by warthog618** [comment]
Looking at the output in a little more detail:

Commands which COULD return more than one object will always return an array of objects, even if only one or no objects are found.

Commands which can only return one object will always return a single standalone object.

Considering what to emit when an object is missing or cannot be accessed.
The returned object could be:
 - an empty object
 - an error object
 - an object containing only the identifying fields provided at the command line
 - ??

To use the same flow for JSON and text, in both cases collecting a set of objects and selecting a formatter post-collection, will require collecting Vec<Result<object, error>>, not Vec<object>, and providing a custom serializer to handle the error case.

e.g. text output with error cases:
```
$ gpiocdev chip
unable to open '/dev/gpiochip0'
$ gpiocdev line apple peach
gpiochip0 5	apple           	output
cannot find line 'peach'
```
The error in the Result must contain the object identifying fields, as well as the root cause of the error.

@fpagliughi Opinions or better suggestions?

**Comment by fpagliughi** [comment]
> Commands which COULD return more than one object will always return an array of objects, even if only one or no objects are found.
>
>Commands which can only return one object will always return a single standalone object.

Yes, perfect.

> Considering what to emit when an object is missing or cannot be accessed.

Interesting. This seems like two possibilities. I think if you search for something against a filter and find zero matches, that should be considered a _successful search returning zero results_. If that makes sense. But any other type of error should return an error message.

To be more specific:
- If you get no results matching the search:
    1. Print and empty object (array?) to stdout
    2. Return a zero (success) from the app/process
- If you receive any other type of error:
    1. Send an error message to _stderr_
    2. Return a non-zero (failure) from the app/process

That's my initial thinking, but I'm not totally married to all the details. The most important thing is that an automated system such as a script that spawns the CLI can tell the difference easily enough.

**Comment by warthog618** [comment]
The cli already behaves as per your specifics, i.e. the usual unix behaviour.
My question is specifically how to map the textual output onto JSON.
(the exit codes will be the same for JSON and text, only the stdout/stderr output would differ)

The bare chip command returns the access state of all chips in the system, and identifies any that are not accessible to the user.
The bare line command returns the list of lines that are accessible to the user, totally ignoring any on inaccessible chips.
Both the base chip and line command return empty lists and success if there are no matches.
i.e. no chips for the chip command or no accessible lines for line command.

For non-bare commands, any command that tries to access a specific chip, which is either missing or has the wrong permissions is an error.  So that applies to lines identified by (chip, offset) as well.

Lines identified by name are handled differently depending on the nature of the operation.
A matching line may or may not exist, or there may be duplicates.
Generally the commands look for the first match and go with that.
If no matches are found, that is an error.
For the line command, where the name is a filter, I would expect that to be an empty list in JSON.
For the other commands, where it is a specifier, I would expect it to indicate an error.

All the commands that take a named line also have the ```--strict``` option which means they don't stop at the first match.
For the line command, all matching lines are returned - so the user can sort out which one they want based on the chip label or something.
For the other commands, like set or get, a check is performed to ensure there is only one match, and of not do not perform the operation and return an error instead.

I'm thinking there has to be an error object in order to provide the same information as the text form.

So a bare ```chip``` would emit ```{ "chips": [{chipinfo},{chipinfo},{error},{chipinfo}...]}```
```chip 3``` would emit ```{chipinfo}``` OR ```{error}```
line is similar - always an array, and elements may be line info or error - ```{"lines":[{lineinfo}, {error}...]```

The other commands that return a single object would return that - or error.
And the ```edges``` and ```notify``` would return ```{event}{event}{event}{event}...``` or ```{error}``` (if they can't setup)
(pardon the simplified JSON - the ```{xxx}``` would be the object xxx serialized to JSON - hopefully you get the idea)

So what should the error look like in JSON?
Do we wrap the error and object in a Result-ish wrapper that contains either the error or the object?
e.g. ```{"info":{lineinfo}``` or ```{"error":{error details}}```

That makes sense to me, though it is a bit more verbose than I would like.
All outputs are either a result or an array of results, JSON encoded.
i.e.
```{result}``` or
```{ "chips": [{result},{result},{result},{result}...]}``` or
```{ "lines": [{result},{result},{result},{result}...]}```

except the event generators of course - those are still a stream of events.

Alternatively, accept that the semantics of the JSON and text outputs are different and omit errors from the JSON output and have the client rely on the exit code.  Not thrilled about that option either.

**Comment by warthog618** [comment]
I also initially overlooked the error cases, and was thinking we'd only emit the appropriate info objects.
But life wasn't meant to be easy.

I do prefer that deserializers are context free though - tends to make streaming simpler - don't have to worry as much about keeping everything in sync.  Not that we are streaming requests over JSON yet....

> I'm not sure if that's necessary. If any one fails, you can just return the error. That might be sufficient.

Wrt the array of chip info, the errors are the chips the user can't access.  If you fail the whole request because you can't access one then you wont be able to see any chips (well unless you try to guess their ids and ask for them singularly).
The other alternative is to just ignore the chips that are inaccessible, as per the bare lines command.
i.e. rework the semantics a little to simplify the JSON - if that suits the use cases.

I'm not too concerned that I haven't spent much effort on how to generate that with serde yet.  I'm sure you could get it to emit whatever you want, one way or another, though that may require the inputs being more complex than you would like.

Anyway, at the moment I've still got a fair bit of code restructuring to get through - to stop those errors being emitted when they are found, and instead propagate them to the top level, where they can be serialised, or not, depending what we decide on.

**Comment by fpagliughi** [comment]
Yeah, in the context of console output from the CLI, if you're specifically building up JSON objects, you can do whatever you want., and build up the objects a field at a time, or even drop in raw JSON objects.

I was thinking ahead about parsing it back in Rust, but maybe that shouldn't be much of a concern for the CLI tool(s) as far as how it combines parts of the output. Whatever gives the most info to the user does seem to make the most sense.

But should the library itself define serializable/deserializable error structs or should that just be put in the CLI?

I was also thinking (possibly way ahead) about how all this serialization might be used in other utilities. I do a lot with [libiio](https://wiki.analog.com/resources/tools-software/linux-software/libiio_internals), and one thing I really like is the network server (iiod) for it... You can write an app for your laptop that interacts with analog I/O as if it were local. That might be cool to have for GPIO as well, and could be built with these serializable objects pretty easily.

**Comment by warthog618** [comment]
> Yeah, in the context of console output from the CLI, if you're specifically building up JSON objects, you can do whatever you want., and build up the objects a field at a time, or even drop in raw JSON objects.
> 

Yeah, this is all CLI specific.  Other tools can do it differently.

> I was thinking ahead about parsing it back in Rust, but maybe that shouldn't be much of a concern for the CLI tool(s) as far as how it combines parts of the output. Whatever gives the most info to the user does seem to make the most sense.
> 

And I was thinking of parsing it in python or jq.

I'm thinking that the ```--json``` mode is more of a fair weather tool, so I'm tempted to simplify the semantics for that mode, so ignore unreadable chips and otherwise return an error where you would return a list containing errors.
That removes the heterogeneous array cases.

But I'm not prepared to select either option at this point.
Once I sort out the code restructuring, I will try some examples with various deserialisation options to see how they fare.
If heterogeneous arrays are easy then might as well keep the semantics as is.
If they are difficult or particularly ugly then simplification is the way to go.

> But should the library itself define serializable/deserializable error structs or should that just be put in the CLI?
> 

The error structs are CLI specific, as the ops are CLI level ops - one level higher than the lib.
Though the lib Error enum should be serialisable - currently it isn't.

> I was also thinking (possibly way ahead) about how all this serialization might be used in other utilities. I do a lot with [libiio](https://wiki.analog.com/resources/tools-software/linux-software/libiio_internals), and one thing I really like is the network server (iiod) for it... You can write an app for your laptop that interacts with analog I/O as if it were local. That might be cool to have for GPIO as well, and could be built with these serializable objects pretty easily.

My stance here is still that the library should support serialisation of the basic types, and the functionality necessary to retrieve them, and that apps can incorporate those in aggregates as they see fit.
Though if there are particular aggregates that see a lot of use then the library could provide those too.
We'll see.

A daemon would be nice.  It is the choice and definition of the API for it that is the sticking point for me.
Given that it would be pretty straight forward to knock one out.

**Comment by fpagliughi** [comment]
Yep. To all of it.

I was thinking that adding it to the CLI was as much an example of serialization as having practical usage. It can evolve over time as needed if it gets some use. I'll be using it, but my use case is simple, and any failure mode would suffice.

> A daemon would be nice. It is the choice and definition of the API for it that is the sticking point for me.

That's the cool thing about `libiio`. The API is the API. One line at the start of the app says if you're connecting "locally" to the hardware or communicating to a remote server over TCP, USB, etc. Everything else is the same.

**Comment by warthog618** [comment]
> That's the cool thing about `libiio`. The API is the API. One line at the start of the app says if you're connecting "locally" to the hardware or communicating to a remote server over TCP, USB, etc. Everything else is the same.

As long as the client-server link is connection oriented that would work.
But the main thing people are asking for from a GPIO daemon is persistence - being able to connect, setup a hardware state, walk away and come back and change things later.  That would involve adding some form of session tracking or authentication.
There is some work being done on a DBus GPIO daemon using libgpiod.  I should take a look at what they are doing.

**Comment by warthog618** [comment]
I've created a json branch that contains where I'm at with --json.  That is complete with the exception of `gpiocdev set` as I'm still considering how best to handle that case.
The general approach I've gone with is not to use heterogeneous arrays, but instead to split the responses into objects and errors, as the errors do not necessarily map to individual command line options, and searching/filtering of the json output should be based on keys not ordering.  The resulting output is a little more verbose than I originally had in mind, but hopefully is clearer.
e.g.
```
$ gpiocdev line --json apple banana pear | jq
{
  "lines": [
    {
      "chip": "gpiochip0",
      "offset": 3,
      "name": "banana",
      "direction": "Input"
    },
    {
      "chip": "gpiochip0",
      "offset": 5,
      "name": "apple",
      "direction": "Input"
    }
  ],
  "errors": [
    "cannot find line 'pear'"
  ]
}
```

The output is normally condensed to a single line - you can always pipe that through `jq` to pretty-print as I have done here.

The `edges` and `notify` subcommands are slightly different.  If setup fails they return an object containing the errors as above. 

```
$ gpiocdev edges --json apple banana pear
{"errors":["cannot find line 'pear'"]}
```

Otherwise they return an event object per line as they occur:

```
$ gpiocdev edges --json apple banana
{"chip":"gpiochip0","name":"apple","timestamp_ns":2644551319210,"kind":"Falling","offset":5,"seqno":1,"lineSeqno":1,"timestamp":"2644.551319210"}
{"chip":"gpiochip0","name":"apple","timestamp_ns":2645056343019,"kind":"Rising","offset":5,"seqno":2,"lineSeqno":2,"timestamp":"2645.56343019"}
{"chip":"gpiochip0","name":"apple","timestamp_ns":2645706095325,"kind":"Falling","offset":5,"seqno":3,"lineSeqno":3,"timestamp":"2645.706095325"}
```

Wrt the `set` subcommand, the only output is in interactive mode, and the open question there is whether to support json input as well as output - assuming json is to be supported in this mode at all.  We are getting into daemon territory here - which might be better served by a dedicated daemon.

**Comment by warthog618** [comment]
Now merged into master.

**Comment by fpagliughi** [comment]
Nice! My laptop is currently in a dozen pieces on the kitchen counter, but if/when I get it back to one piece I will try this out.

**Comment by warthog618** [comment]
Released in cli v0.5.0.
